contributingGuides/TYPESCRIPT_STYLE.md

```diff
@@ -0,0 +1,886 @@
+
+```ts
+// Before
+const propTypes = {
+    unknownData: PropTypes.any,
+    anotherUnknownData: PropTypes.any,
+    indexes: PropTypes.arrayOf(PropTypes.number),
+    items: PropTypes.arrayOf(PropTypes.shape({
+        value: PropTypes.string,
+        label: PropTypes.string,
+    })),
+    shouldShowIcon: PropTypes.bool,
+    onChangeText: PropTypes.func,
+    count: PropTypes.number,
+    session: PropTypes.shape({
+        authToken: PropTypes.string,
+        accountID: PropTypes.number,
+    }),
+    errors: PropTypes.objectOf(PropTypes.string),
+    inputs: PropTypes.objectOf(PropTypes.shape({
+        id: PropTypes.string,
+        label: PropTypes.string,
+    })),
+    label: PropTypes.string,
+    anchor: PropTypes.node,
+    footer: PropTypes.element,
+    uniqSymbol: PropTypes.symbol,
+    icon: PropTypes.elementType,
+    date: PropTypes.instanceOf(Date),
+    size: PropTypes.oneOf(['small', 'medium', 'large']),
+};
+
+// After
+type Item = {
+    value: string;
+    label: string;
+};
+
+type Session = {
+    authToken: string;
+    accountID: number;
+};
+
+type Input = {
+    id: string;
+    label: string;
+};
+
+type Size = 'small' | 'medium' | 'large';
+
+type Props = {
+    unknownData: string[];
+
+    // It's not possible to infer the data as it can be anything because of reasons X, Y and Z.
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    anotherUnknownData: any;
+
+    indexes: number[];
+    items: Item[];
+    shouldShowIcon: boolean;
+    onChangeText: (value: string) => void;
+    count: number;
+    session: Session;
+    errors: Record<string, string>;
+    inputs: Record<string, Input>;
+    label: string;
+    anchor: React.ReactNode;
+    footer: React.ReactElement;
+    uniqSymbol: symbol;
+    icon: React.ElementType;
+    date: Date;
+    size: Size;
+};
+```
+
+Please note that you may encounter some PropTypes marked as `isRequired`, but when migrating to `type` you shall first consider all PropTypes as required ones and then inspect the component's source code and usage to identify which ones are really required or optional.
+
+```ts
+// Before
+const propTypes = {
+    isVisible: PropTypes.bool.isRequired,
+    // `confirmText` prop is not marked as required here, theoretically it is optional.
+    confirmText: PropTypes.string,
+};
+
+// After
+type Props = {
+    isVisible: boolean;
+    // Consider it as required unless you have proof that it is indeed an optional prop.
+    confirmText: string;
+};
+```
+
+### Default props
+
+As **defaultProps** are not allowed anymore in the project, use object destructuring in your `Props` in order to set default values to them. The props that have default value must also be marked as optional in your `Props` type.
+
+```tsx
+// Before
+const propTypes = {
+    requiredProp: PropTypes.string.isRequired,
+    optionalPropWithDefaultValue: PropTypes.number,
+    optionalProp: PropTypes.bool,
```

Comment:
I think we don't currently allow optional props without default values, according to our style guide. Maybe I'm wrong, but why should we allow optional props without default values?

Reply:
Hi @neil-marcellini , could you point me out where do we have this specification in our style guides? I didn't find it.

About the optional props, the reason is because we may have a situation where the prop of that component is optional and don't need to or can't have a default value set. For example:

```tsx
import React from 'react';
import {StyleProp, Text, TextStyle} from 'react-native';

type Props = {
  // The developer can apply a custom styling to this component
  // but it doesn't make sense to have a default value, so when using
  // the prop it will be `undefined` unless set.
  style?: StyleProp<TextStyle>;

  // Required prop.
  label: string;

  // The developer can set a custom count, but we need a value
  // anyway so we are going to set `0` as a default value.
  count?: number;
};

function LabelWithBadge({style, label, count = 0}: Props) {
  return (
    <Text style={[{color: 'red'}, style]}>{`${label} - Count: ${count}`}</Text>
  );
}

function App() {
  return <LabelWithBadge label="Notifications" />;
}
```

Reply:
Ok got it thank you for the example. Yeah it's not mentioned in our App style guide. It is mentioned in the Airbnb javascript/react style guide, which was are supposed to follow, in the [props section](https://github.com/airbnb/javascript/tree/master/react#props).

> Always define explicit defaultProps for all non-required props.

In the example you showed above we would provide a default of `null` for the optional prop `style`. I think that's helpful, but it's not critical. Maybe we can get some more opinions. @hayata-suenaga @roryabraham

Reply:
@neil-marcellini

> I think we don't currently allow optional props without default values

thank you for reminding us of this 👍 

Are we discussing the necessity of providing default values for optional properties?

```ts
type MyCompProps = { style?: StyleProp<TextStyle>; };

function MyComp({ style }: MyCompProps) {...}
// vs.
function MyComp({ style = null }: MyCompProps) {...}
```

IMO, it depends. We don't have to always require default values. If a value's default value is `undefined` (or `null`), it might be redundant to explicitly assign `undefined` to it.

Or if it's easy to handle an optional prop having `undefined` value inside a component, it's also not necessary to come up with default values for these.

```tsx
type MyCompProps = { person?: { name: string } };

function MyComp({ person }: MyCompProps} {
   ...<Text>{person?.name ?? 'anonymous'}</Text>...
}
```

Reply:
> We don't have to always require default values. If a value's default value is undefined (or null), it might be redundant to explicitly assign undefined to it.

Totally agree 👍 

My take on this is that with Typescript we don't need to always provide default props, since we can rely on type checking to identify any missing props. However, it's important to remember that having default values can be helpful in some cases. This way, we don't need to check if a prop exists every time it's used. The goal is to find a good balance between readable code, simplicity and safety.

Reply:
I agree that we don't need to always have a default since TypeScript will warn us if a prop is possibly undefined 👍🏼